common/src/main/java/org/apache/nemo/common/DirectByteBufferInputStream.java

```diff
@@ -0,0 +1,71 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.nemo.common;
+
+import java.io.EOFException;
+import java.io.IOException;
+import java.io.InputStream;
+import java.nio.ByteBuffer;
+import java.util.List;
+
+/**
+ * This class is a customized input stream implementation which reads data from
+ * list of {@link ByteBuffer}.
+ */
+public class DirectByteBufferInputStream extends InputStream {
+  private List<ByteBuffer> bufList;
+  private int current = 0;
+
+  /**
+   * Default Constructor.
+   *
+   * @param bufList is the target data to read.
+   */
+  public DirectByteBufferInputStream(final List<ByteBuffer> bufList) {
+   this.bufList = bufList;
+  }
+
+  /**
+   * Reads data from the list of {@code ByteBuffer}s.
+   *
+   * @return integer.
+   * @throws IOException
+   */
+  @Override
+  public int read() throws IOException {
+    return getBuffer().get() & 0xff;
```

Comment:
You can probably refer to this website: https://rules.sonarsource.com/java/RSPEC-3034 . As far as I know, this is a java-specific issue. Since the raw byte value in java represents signed value,  it should be masked to get the proper value. I will mention this in the comment thanks :)